modules/imgcodecs/include/opencv2/imgcodecs.hpp

```diff
@@ -332,6 +332,51 @@ CV_EXPORTS_W bool haveImageReader( const String& filename );
  */
 CV_EXPORTS_W bool haveImageWriter( const String& filename );
 
+/** @brief To read Multi Page images on demand
+
+The ImageCollection class provides iterator API to read multi page images on demand. Create iterator
+to the collection of the images and iterate over the collection. Decode the necessary page with operator*.
+
+The performance of page decoding is O(1) if collection is increment sequentially. If the user wants to access random page,
+then the time Complexity is O(n) because the collection has to be reinitialized every time in order to go to the correct page.
+However, the intermediate pages are not decoded during the process, so typically it's quite fast.
+This is required because multipage codecs does not support going backwards.
+After decoding the one page, it is stored inside the collection cache. Hence, trying to get Mat object from already decoded page is O(1).
+If you need memory, you can use .releaseCache() method to release cached index.
+The space complexity is O(n) if all pages are decoded into memory. The user is able to decode and release images on demand.
+*/
+class CV_EXPORTS ImageCollection {
+public:
+    struct CV_EXPORTS iterator {
+        iterator(ImageCollection* col);
+        iterator(ImageCollection* col, int end);
+        Mat& operator*();
+        Mat* operator->();
+        iterator& operator++();
+        iterator operator++(int);
+        friend bool operator== (const iterator& a, const iterator& b) { return a.m_curr == b.m_curr; };
+        friend bool operator!= (const iterator& a, const iterator& b) { return a.m_curr != b.m_curr; };
+
+    private:
+        ImageCollection* m_pCollection;
+        int m_curr;
+    };
+
+    ImageCollection();
+    ImageCollection(const String& filename, int flags);
+    void init(const String& img, int flags);
+    size_t size() const;
+    Mat& at(int index);
+    Mat& operator[](int index);
```

Comment:
Reference as return value destroys the idea with cache. `resetCache` calls `release` for all images, including the image user has reference to. It's not what expected. Previous version relied on reference counting inside of `cv::Mat` and worked well. I propose to restore it back.

Reply:
Actually, `releaseCache` only calls `release` method on the given index page. For example, suppose you decode all pages into cache,

```cpp
ImageCollection col("filename.txt");
for(auto& i: coll){
     *i; // Decode all pages so they can go to cache

coll.releaseCache(2); // only release page 3 from the cache. 
auto mat = coll.at(1); // No decoding happens because page 2 is still at the cache
```

With returning reference, now it is user responsibility to not call `releaseCache` on the index that he is referring to. 

What do you think? And why do you think `releaseCache` releases for all images? I have never mentioned that and noone gave me feedback about that. Should I make that method to release for all images?

Reply:
I rechecked all cases and it looks like the only dangerous case is something like this:
```
auto current_frame = decoder.at(i);
decoder.releaseCache();
```
releaseCache invalidates references (`auto` is `Mat&`), but it's expected behavior. In all other cases like `cv::Mat img = decoder.at(i);` `cv::Mat` copy constructor is called and user code does not suffer. So you are right! Let's go ahead with references then.

Reply:
In your example, `auto` is deduced to `Mat`. If user wants to get reference to Mat object, he needs `auto&`. Anyway, it is still safe. Thanks!